contracts/exchangeIssuance/ExchangeIssuanceZeroEx.sol

```diff
@@ -0,0 +1,561 @@
+    }
+
+    /**
+     * Redeems a given list of SetToken components for given token.
+     *
+     * @param _setToken             The set token being swapped.
+     * @param _amountSetToken       The amount of set token being swapped.
+     * @param _swaps                An array containing ZeroExSwap swaps.
+     * @param _outputToken          The token for which to sell the index components
+     *
+     * @return totalOutputTokenBought  Total amount of output token received after liquidating all SetToken components
+     */
+    function _sellComponentsForOutputToken(ISetToken _setToken, uint256 _amountSetToken, bytes[] memory _swaps, IERC20 _outputToken, address _issuanceModule)
+        internal
+        returns (uint256 totalOutputTokenBought)
+    {
+        (address[] memory components, uint256[] memory componentUnits) = getRequiredRedemptionComponents(_issuanceModule, _setToken, _amountSetToken);
+        uint256 outputTokenBalanceBefore = _outputToken.balanceOf(address(this));
+        for (uint256 i = 0; i < _swaps.length; i++) {
+            uint256 maxAmountSell = componentUnits[i];
+
+            uint256 outputTokenAmountBought;
+            uint256 componentAmountSold;
+
+            // If the component is equal to the input token we don't have to trade
+            if(components[i] == address(_outputToken)) {
+                totalOutputTokenBought = totalOutputTokenBought.add(maxAmountSell);
+                componentAmountSold = maxAmountSell;
+            }
+            else {
+                _safeApprove(IERC20(components[i]), address(swapTarget), maxAmountSell);
+                uint256 componentBalanceBefore = IERC20(components[i]).balanceOf(address(this));
+                _fillQuote(_swaps[i]);
+                uint256 componentBalanceAfter = IERC20(components[i]).balanceOf(address(this));
+                componentAmountSold = componentBalanceBefore.sub(componentBalanceAfter);
+                require(maxAmountSell >= componentAmountSold, "ExchangeIssuance: OVERSOLD COMPONENT");
+            }
+
+        }
+        uint256 outputTokenBalanceAfter = _outputToken.balanceOf(address(this));
+        totalOutputTokenBought = totalOutputTokenBought.add(outputTokenBalanceAfter.sub(outputTokenBalanceBefore));
+    }
+
+    /**
+     * Execute a 0x Swap quote
+     *
+     * @param _quote          Swap quote as returned by 0x API
+     *
+     */
+    function _fillQuote(
+        bytes memory _quote
+    )
+        internal
+        
+    {
+
+        (bool success, bytes memory returndata) = swapTarget.call(_quote);
+
+        // Forwarding errors including new custom errors
+        // Taken from: https://ethereum.stackexchange.com/a/111187/73805
+        if (!success) {
+            if (returndata.length == 0) revert();
+            assembly {
+                revert(add(32, returndata), mload(returndata))
+            }
```

Comment:
Works! (Saw a `UniswapFeature/UnderBought` error while running the integration tests)

Reply:
Glad to hear that. Will give the SO where I took this from an upvote for this. 😅
I previously tried adding a test contract that returns a solidity `CustomError` for testing this in the unit tests but gave up on it after a day since the necessary hardhat / ethers upgrades to support solidity 0.8.4 got me into all kinds of trouble.